Transaction support

tub opened the pull request:
This PR adds transaction support to ZK copy, allowing the user to specify the number of operations in the transaction with -b/--batchSize.

We've tested this with up to 10,000 operations per transaction with 100,000 nodes and seen around a 6x write speedup.

Another optimization for users wanting to perform a copy multiple times is the -m/--mtime flag, which allows zkcopy to ignore nodes older than the specified mtime, skipping any writes that may have been performed by a previous run of zkcopy. To aid this, zkcopy logs the maximum mtime of the copied nodes at runtime.

This PR also refactors the CLI arg parsing to use picocil, removing a lot of manual parsing and type-checking.

tub committed:
Batch operations in to transactions for faster writes

tub committed:
Allows override of transaction size, refactors cli parsing

Replacing hand-rolled cli arg parsing with picocli library.

tub committed:
Updating readme for new usage/help

tub committed:
Showing default values in help

tub committed:
Adding -h --help

tub committed:
Lowering default transaction size, adding more help

kshchepanovskyi commented:
Thanks, looks awesome!
I'd like to write few tests for current version before merging PR to make sure nothing is broken, using https://www.testcontainers.org. Probably will do it on a weekend.

Commit:
Bumped session timeout and add explicit close of Zookeeper object to avoid leaking sessions

Commit:
Updates

tub committed:
Allowing override of ZK session timeout, adds tests

The default session timeout is now 40s, as this tool
will generally create fairly large transactions which
need a larger timeout.

This change also refactors a few classes to make
them easier to test, e.g by moving the ZooKeeper object
to a parameter rather than being constructed at a lower level.

tub committed:
Close Zookeeper connection after writing

tub committed:
Merge branch 'add-session-timeout-and-close-1' into transaction-support

tub committed:
Passing down sessionTimeout into reader threads

tub committed:
Updating readme for --timeout param

tub commented:
I've added a bunch of JUnit unit-tests around the code we've changed. Not yet looked into the testcontainers setup, but it looks interesting.
The unit tests required a bit of refactoring in order to pass dependencies in.

We (myself and @ronin13) have also worked on cleaning up the zookeeper server object once zkcopy is finished, and raised the session timeout so that large transactions and reads don't fail.

tub committed:
Closes the reader zookeeper connections manually

Previous implementation used finalize() which would
not always be called. This solution manually calls
close once all reader threads have completed.

tub committed:
Adds initial testcontainers.org integration test

To run the integration test just call `mvn verify`
make sure that you are running in an environment with
a docker daemon available. It should be automatically
discovered by testcontainers.

tub commented:
I've just added an initial testcontainers integration test. Really nice library, thanks for the headsup @kshchepanovskyi 
The new itest highlighted that the connections to the source ZK were still not being closed so I've scrapped the finalize() implementation and exposed a way of closing the ZooKeepers from the ThreadFactory.

tub committed:
Fixing docker build to not try to use docker in docker

kshchepanovskyi commented on README.md:
```diff
@@ -27,16 +27,34 @@ docker run --rm -it ksprojects/zkcopy --source server:port/path --target server:
+                                Default: -1
+  -s, --source=server:port/path
+                              location of a source tree to copy
+                                Default: 10-81-26-99-uswest2cdevc:
```
As far as I see there's no default value in the code (and there should not be one), please update readme.

tub replied:
Good spot, I think that must have snuck in when we generated the help with some auto-defaults in there.

tub committed:
Removes spurious defaults from readme
